### ISSUE TITLE
"follows"-behavior changes based on existence of lockfile

### ISSUE THREAD

**Issue by queezle42** [NixOS/nix#4687]
**Describe the bug**

I have three flakes (a, b, c) that are set up as follows:
```
❯ nix flake metadata --no-write-lock-file 'gitlab:jens/flake-follows-bug-poc/with-lockfile-b?host=git.c3pb.de&dir=a'
[...]
Inputs:
├───b: gitlab:jens/flake-follows-bug-poc/fe4008332af24b8258363d29cbc1e309b16b7a2d?dir=b
│   ├───c: gitlab:jens/flake-follows-bug-poc/97302a222ab4ec7f93fa9b304e1661cee1b5f01f?dir=c
│   │   └───nixpkgs follows input 'nixpkgs'
│   └───nixpkgs follows input 'nixpkgs'
└───nixpkgs: github:nixos/nixpkgs/04a2b269d8921505a2969fc9ec25c1f517f2b307
```

When I remove the lockfile of flake "b" the "follows"-chain for `a/b/c/nixpkgs` breaks:
```
❯ nix flake metadata --no-write-lock-file 'gitlab:jens/flake-follows-bug-poc/without-lockfiles?host=git.c3pb.de&dir=a'
[...]
Inputs:
├───b: gitlab:jens/flake-follows-bug-poc/a48ccc743f51598ae54c8cc8a2f6c2f83ebafadb?dir=b
│   ├───c: gitlab:jens/flake-follows-bug-poc/a48ccc743f51598ae54c8cc8a2f6c2f83ebafadb?dir=c
│   │   └───nixpkgs: github:nixos/nixpkgs/eef00d30461a4624daf1089e8f8a3fb2ec7da0af
│   └───nixpkgs follows input 'nixpkgs'
└───nixpkgs: github:nixos/nixpkgs/04a2b269d8921505a2969fc9ec25c1f517f2b307
```

**Steps To Reproduce**

POC source code can be found at https://git.c3pb.de/jens/flake-follows-bug-poc/-/tree/without-lockfiles and https://git.c3pb.de/jens/flake-follows-bug-poc/-/tree/with-lockfile-b, commands are included above.

**Expected behavior**

I would expect that the `follows`-attribute is honored even if no lockfile exists.

**`nix-env --version` output**
`nix-env (Nix) 2.4pre20210326_dd77f71`

**Comment by queezle42** [comment]
I've tested this with the current `master` and the behavior hasn't changed.

**Comment by dvzubarev** [comment]
Hi, thank you for this repositories. I've stumbled upon confusing behavior of transitive follows and can reproduce on your data.
I'm using Nix 2.5.1
Commands:

```bash
git clone 'https://git.c3pb.de/jens/flake-follows-bug-poc' -b with-lockfile-b
cd flake-follows-bug-poc/a/
nix flake update .
```
For now its ok, only one nixpkgs in flake.lock as it should be.
But run any command and flake.lock is overridden... 

```bash
nix flake metadata .

warning: Git tree '/tmp/follows/flake-follows-bug-poc' is dirty
warning: updating lock file '/tmp/follows/flake-follows-bug-poc/a/flake.lock':
• Updated input 'b/c/nixpkgs':
    follows 'b/nixpkgs'
  → 'github:nixos/nixpkgs/9789ed45b759c8f770bcba4b06c1a1c5ec386664' (2022-01-18)
warning: Git tree '/tmp/follows/flake-follows-bug-poc' is dirty
Resolved URL:  git+file:///tmp/follows/flake-follows-bug-poc?dir=a
Locked URL:    git+file:///tmp/follows/flake-follows-bug-poc?dir=a
Path:          /nix/store/p8b04ma0ghd8s580hmymp9v5s5ns3lv7-source
Last modified: 2021-04-03 05:47:38
Inputs:
├───b: gitlab:jens/flake-follows-bug-poc/fe4008332af24b8258363d29cbc1e309b16b7a2d?dir=b
│   ├───c: gitlab:jens/flake-follows-bug-poc/97302a222ab4ec7f93fa9b304e1661cee1b5f01f?dir=c
│   │   └───nixpkgs: github:nixos/nixpkgs/9789ed45b759c8f770bcba4b06c1a1c5ec386664
│   └───nixpkgs follows input 'nixpkgs'
└───nixpkgs: github:nixos/nixpkgs/5aaed40d22f0d9376330b6fa413223435ad6fee5
```
Now there are two nixpkgs in lockfile...

Then run flake update to overwrite lockfile again...
```bash
nix flake update .
warning: Git tree '/tmp/follows/flake-follows-bug-poc' is dirty
warning: updating lock file '/tmp/follows/flake-follows-bug-poc/a/flake.lock':
• Updated input 'b/c/nixpkgs':
    'github:nixos/nixpkgs/9789ed45b759c8f770bcba4b06c1a1c5ec386664' (2022-01-18)
  → follows 'b/nixpkgs'
warning: Git tree '/tmp/follows/flake-follows-bug-poc' is dirty
```

**Comment by dvzubarev** [comment]
It seems to be fixed in 2.6.0.
I can't reproduce confusing behavior from my  [comment](https://github.com/NixOS/nix/issues/4687#issuecomment-1015179364) anymore. Maybe your issue was fixed too.

**Comment by queezle42** [comment]
Yes, I think this issue is resolved. Thanks for notifying!
